### DISCUSSION TITLE
Native binary uuid support

### DISCUSSION THREAD

**Discussion by denny-bee** [laravel/framework#51605] [Ideas]
Many people out there recommend storing uuids as binary. There are some archived packages which are trying to workaround and are not optimal, because they seem to work, but come with many other problems.
Maybe there could be a way to declare two extra methods for custom casts which are used for writing and retrieving data from the database. 
This native handling would be way better then some workarounds of the orm.

**Comment by plumthedev** [comment]
It will be available in the next Laravel 12 release

**Reply by marius-ciclistu** [reply]
@plumthedev 
With your PR https://github.com/laravel/framework/pull/58254/files I guess the type of the PK should be string. This will not interfere with eloquent's logic of auto adding cast to int for pks.

As I saw, casting in eloquent introduces delays and many [set](https://github.com/laravel/framework/discussions/31778) calls. Do you know if the gains of using binary pks exceeds the losses from the casting system?

**Reply by marius-ciclistu** [reply]
Also, if you need to perform a query like

```php

Model::query()->where('fpk', $bynaryPkModel->id)->first();
```

You would need to manually cast to bynary the id.

This means that all tables using that binary must follow the same cast procedure. Valid also for relations.

**Reply by plumthedev** [reply]
1. for sure the binary columns will be much faster, even if set will be called N times, I will provide the benchmarks soon

2. QB support will be added soon, you can provide an PR if you want, I thought about the methods with binary prefix 'whereBinary/whereBinaryIn' and so

**Reply by marius-ciclistu** [reply]
For me bulletproof auto handling the queries seems mission impossible...

**Reply by marius-ciclistu** [reply]
There was a similar need for handling all the orWhere to prevent data leaks. Not accomplished.

The new methods don't solve the relations problems.

**Reply by marius-ciclistu** [reply]
What do you think of using binaries in php and convert them to strings only when the data leaves on a http response or log for example.

That usually happens via toArray method for models.

When you create model with some values, then you could convert from string to uuid if needed.

This would not break the relations. 

[Mutator](https://laravel.com/docs/12.x/eloquent-mutators#defining-a-mutator)  could be used to set the binary from string if it is string.

But this will mean reading the data from model will return binary not string as it would in toArray which will cause logs or views to contain binary for example.

Maybe new methods on the model to read and convert to string would help like getBinaryAttrributeValueAsString.

From any side you look at this it will break something. Casts break relations. This idea breaks logs , views etc.

**Reply by plumthedev** [reply]
> For me bulletproof auto handling the queries seems mission impossible...

Nobody is talking about auto-handling the queries, but even if so, during the work with Eloquent QB we can detect if the column value should be casted to binary, as we have model instance provided. For the standard QB, we will add new methods to encode/decode the binary value. And in that case - yes, developer should explicity use the binary method to support it. 

> The new methods don't solve the relations problems.

QB will.

> What do you think of using binaries in php and convert them to strings only when the data leaves on a http response or log for example.

This is exactly how casting will work

**Reply by macropay-solutions** [reply]
@plumthedev

>Nobody is talking about auto-handling the queries, but even if so, during the work with Eloquent QB we can detect if the column value should be casted to binary, as we have model instance provided. For the standard QB, we will add new methods to encode/decode the binary value. And in that case - yes, developer should explicitly use the binary method to support it.

 We did something similar using column closures when we transform the request into SQL via Eloquent Query Builder (for hashes for example) but it could work also for binary conversion. The thing is that we do that when we translate the request array into Eloquent Query Builder / Query Builder. That is a level up from the point that you refer to. We still use the model (including the related model) to retrieve those closures and apply them when we call methods from the builder.
Now you want to put analog logic into the "where" methods, so a level down.

That will be interesting to see because you have the model only in the builder/relation but not in the query builder, and some methods are forwarded to the QB from EQB.

>This is exactly how casting will work

Casting does not work like that.

Once the model is retrieved from DB and hydrated, if you access id, the id will be string. Analog if you do toArray. But when you load a relation, the same string will be used in the query and it will not be converted into binary for the where clause.  For the save logic, yes, it will convert to binary.

**Comment by plumthedev** [comment]
```
======================================================================
UUID/ULID BENCHMARK WITH JOINS - MySQL 8 OPTIMIZED
======================================================================

Connecting to MySQL...

Applying MySQL 8 optimizations (may require privileges)...
  Applied: SET GLOBAL innodb_buffer_pool_size = 1073741824...
  Skipped (needs SUPER): SET GLOBAL innodb_log_file_size = 268435... (1238 (HY000): Variable 'innodb_log_file_size' is a read only variable)
  Applied: SET GLOBAL innodb_flush_log_at_trx_commit = 2...
  Applied: SET SESSION innodb_strict_mode = ON...
Tables created successfully
Generating test data for 66,666 users...
  Generated data for 50,000 users...
  Total: 66,666 users, 367,014 orders, 1,101,582 items

======================================================================
INSERT BENCHMARK
======================================================================
All tables truncated
UUIDV4_VARCHAR          26.68s
UUIDV4_BINARY           18.97s
UUIDV7_VARCHAR          26.20s
UUIDV7_BINARY           14.99s
ULID_VARCHAR            27.24s
ULID_BINARY             15.06s

======================================================================
JOIN BENCHMARK (5,000 queries each)
======================================================================
UUIDV4_VARCHAR           3.40s
UUIDV4_BINARY            3.00s
UUIDV7_VARCHAR           3.24s
UUIDV7_BINARY            2.82s
ULID_VARCHAR             3.30s
ULID_BINARY              2.96s

======================================================================
AGGREGATION JOIN BENCHMARK (total sales per user, top 1000)
======================================================================
UUIDV4_VARCHAR          30.48s
UUIDV4_BINARY           18.92s
UUIDV7_VARCHAR          32.94s
UUIDV7_BINARY           15.99s
ULID_VARCHAR            26.79s
ULID_BINARY             16.41s

======================================================================
RANGE JOIN BENCHMARK (recent completed orders with items)
======================================================================
UUIDV4_VARCHAR          18.83s
UUIDV4_BINARY           12.62s
UUIDV7_VARCHAR          24.46s
UUIDV7_BINARY            9.54s
ULID_VARCHAR            28.34s
ULID_BINARY              9.70s

======================================================================
TABLE & INDEX SIZES (combined users + orders + items)
======================================================================
Type                      Data MB     Index MB     Total MB
----------------------------------------------------------
UUIDV4_VARCHAR             279.28       264.58       543.86
UUIDV4_BINARY              172.86       128.94       301.80
UUIDV7_VARCHAR             322.27       280.62       602.89
UUIDV7_BINARY              202.91       138.91       341.81
ULID_VARCHAR               268.20       215.34       483.55
ULID_BINARY                201.89       137.92       339.81

======================================================================
FINAL SUMMARY
======================================================================

Type                  Insert      Join       Agg     Range   Size MB
--------------------------------------------------------------------
uuidv4_varchar         26.68      3.40     30.48     18.83     543.9
uuidv4_binary          18.97      3.00     18.92     12.62     301.8
uuidv7_varchar         26.20      3.24     32.94     24.46     602.9
uuidv7_binary          14.99      2.82     15.99      9.54     341.8
ulid_varchar           27.24      3.30     26.79     28.34     483.5
ulid_binary            15.06      2.96     16.41      9.70     339.8

======================================================================
WINNERS (lower is better)
======================================================================
Fastest INSERT:           uuidv7_binary
Fastest JOIN:             uuidv7_binary
Fastest AGGREGATION:      uuidv7_binary
Fastest RANGE SCAN:       uuidv7_binary
Smallest SIZE:            uuidv4_binary

======================================================================
OVERALL SCORE (lower is better, weighted: insert=2, join=3, agg=2, range=2, size=1)
======================================================================
uuidv7_binary        0.133
ulid_binary          0.945
uuidv4_binary        2.237
ulid_varchar         8.353
uuidv4_varchar       8.405
uuidv7_varchar       8.593

Benchmark complete!
```

**Reply by marius-ciclistu** [reply]
- A mutator using your encoder
- A getDecodedAttributeValue new method that uses your decoder
- A column closure to transform the value from request in binary on filter
- toArray adjusted to decode the binary into string
-  without casts

These  would not break relations and would allow logs etc.

This means getAttributeValue will return binary so using where('fk', $model->id) will work.

**Reply by marius-ciclistu** [reply]
I forgot about **find** from the get route. That should encode to binary the string from request...

**Comment by plumthedev** [comment]
With the foundational pieces now in place, we can move toward full first-class support for binary primary and foreign keys; the remaining items can be summarized as actionable gaps, each of which can be independently implemented to achieve a truly end-to-end string-based developer experience where all binary conversion is handled transparently by the framework.


* **Eloquent-level transparency for binary keys**: Eloquent should automatically convert string UUID/ULID values to binary form in all core operations, including `find`, `where`, route model binding, and relationship queries, without requiring manual casting by the developer.

* **First-class support for binary keys in relationships**: Eloquent relations (including eager loading) must correctly handle binary primary and foreign keys, ensuring joins and `WHERE IN` clauses operate on binary values internally while exposing string-based APIs externally.

* **Explicit binary key semantics in models**: Laravel should introduce a native way to declare binary primary keys (e.g. `$keyType = 'binary'` or equivalent), allowing the ORM to treat such keys differently from textual strings during query construction.

* **Query Builder helpers for binary identifiers**: The Query Builder and Eloquent should provide dedicated methods (e.g. `whereBinary`, `whereInBinary`, binary-aware `whereKey`) that accept string identifiers and handle binary encoding internally.

* **Schema Blueprint abstractions for binary UUIDs**: Migrations should offer explicit helpers like `binaryUuid`, `binaryUlid`, and matching foreign key shortcuts, creating consistent `BINARY(16)` columns across supported databases without manual configuration.

* **Parity with existing UUID/ULID tooling**: For every text-based UUID/ULID feature (traits, schema helpers, foreign keys, morphs), Laravel should provide an equivalent binary-focused API to ensure consistency and ergonomics.

* **End-to-end string-based developer experience**: Across models, queries, relations, and serialization, developers should work exclusively with readable string identifiers, with all binary conversion handled transparently by the framework.

* **Comprehensive documentation for binary UUID support**: Laravel’s documentation should clearly describe how to define binary UUID/ULID columns in migrations, configure models and casts, work with queries and relationships, and understand the expected string-based developer workflow, so that binary identifiers can be adopted confidently without relying on source code or external packages.

**Reply by marius-ciclistu** [reply]
This reminds me of the php true async RFC in terms of complexity and changes needed. I hope it succeeds without breaking Eloquent.

**Reply by shaedrich** [reply]
Would be a very complete feature, but I wouldn't bank on Taylor rubberstamping all of it
